tests/integration/test_application.py

```diff
@@ -143,6 +144,20 @@ async def test_upgrade_charm_switch(event_loop):
         assert app.data['charm-url'] == 'cs:ubuntu-8'
 
 
+@base.bootstrapped
+@pytest.mark.asyncio
+async def test_upgrade_local_charm(event_loop):
+    async with base.CleanModel() as model:
+        tests_dir = Path(__file__).absolute().parent
+        charm_path = tests_dir / 'upgrade-charm'
+        app = await model.deploy('ubuntu-0')
+        await model.block_until(lambda: (len(app.units) > 0 and
+                                         app.units[0].machine))
+        assert app.data['charm-url'] == 'cs:ubuntu-0'
+        await app.upgrade_charm(path=charm_path)
+        assert app.data['charm-url'] == 'local:ubuntu'
```

Comment:
When I ran this test, I got:

```
juju.errors.JujuAPIError: cannot upgrade application "ubuntu" to charm "local:focal/ubuntu-0": cannot change an application's series
```

Also, this doesn't actually wait for the charm to deploy or upgrade to confirm that it completes successfully.  My suggestion:

```suggestion
        app = await model.deploy('ubuntu', series='focal')
        await model.wait_for_idle(wait_for_active=True)
        assert app.data['charm-url'].startswith('cs:ubuntu')
        await app.upgrade_charm(path=charm_path)
        await model.wait_for_idle()  # nb: can't use wait_for_active because test charm goes to "waiting"
        assert app.data['charm-url'] == 'local:focal/ubuntu-0'
```